ledger/forest/light_forest.go

```diff
@@ -0,0 +1,131 @@
+package forest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/optakt/flow-dps/ledger/trie"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type LightForest struct {
+	Nodes []*trie.LightNode
+	Tries []*trie.LightTrie
+}
+
+func FlattenForest(f *Forest) (*LightForest, error) {
+	tries, _ := f.GetTries()
+	lightTries := make([]*trie.LightTrie, 0, len(tries))
+	lightNodes := []*trie.LightNode{nil} // First element needs to be nil.
+
+	index := make(trie.IndexMap)
+	index[nil] = 0
+
+	count := uint64(1)
+	for _, t := range tries {
+		for itr := trie.NewNodeIterator(t); itr.Next(); {
+			node := itr.Value()
+
+			_, exists := index[node]
+			if exists {
+				continue
+			}
+
+			index[node] = count
+			count++
+
+			lightNode, err := trie.ToLightNode(node, index)
+			if err != nil {
+				return nil, fmt.Errorf("could not build light node: %w", err)
+			}
+			lightNodes = append(lightNodes, lightNode)
+		}
+
+		lightTrie, err := trie.ToLightTrie(t, index)
+		if err != nil {
+			return nil, fmt.Errorf("could not build light trie: %w", err)
+		}
+		lightTries = append(lightTries, lightTrie)
+	}
+
+	lightForest := LightForest{
+		Nodes: lightNodes,
+		Tries: lightTries,
+	}
+
+	return &lightForest, nil
+}
+
+func RebuildTries(store dps.Store, lightForest *LightForest) ([]*trie.Trie, error) {
+	tries := make([]*trie.Trie, 0, len(lightForest.Tries))
+
+	// At this call, we give a 150 million + slice of light nodes and expect 150 million + normal nodes as the return value. As long as the `lightForest` is not garbage collected, memory usage remains huge.
```

Comment:
I should have made it a FIXME because it was more intended as a note to myself than smth that should remain in the code. I will get rid of it and replace it with a note just saying that this is the point at which we are likely to peak in terms of memory usage.